From b56915cf64a7eae9633c8e0b11b34259ed7c7f30 Mon Sep 17 00:00:00 2001
From: Rabeeh Khoury <rabeeh@solid-run.com>
Date: Sun, 1 May 2022 17:06:24 +0300
Subject: [PATCH] lx2160acex7: add 8-x-x and half twins device tree

1. Add 8-x-x SERDES configuration device tree. Mainly used for ClearFog
CX/HoneyComb with the default 8/5/2 SERDES configuration.
2. Add new machine half-twins (SERDES configuration 8/9/2) - tota 8x10G
and 8x1G with PCIe connected to OCP card

Signed-off-by: Rabeeh Khoury <rabeeh@solid-run.com>
---
 arch/arm/dts/Makefile                         |   2 +
 arch/arm/dts/fsl-lx2160a-cex7-8-x-x.dts       | 196 ++++++++++++++++
 arch/arm/dts/fsl-lx2160a-half-twins-8-9-x.dts | 215 ++++++++++++++++++
 board/solidrun/lx2160a/eth_lx2160acex7.c      |  27 ++-
 configs/lx2160acex7_tfa_defconfig             |   2 +-
 5 files changed, 440 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm/dts/fsl-lx2160a-cex7-8-x-x.dts
 create mode 100644 arch/arm/dts/fsl-lx2160a-half-twins-8-9-x.dts

diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 861346835..b084ee0fa 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -426,6 +426,8 @@ dtb-$(CONFIG_FSL_LSCH3) += fsl-ls2080a-qds.dtb \
 	fsl-lx2162a-qds-18-x.dtb\
 	fsl-lx2162a-qds-20-x.dtb\
 	fsl-lx2160a-cex7.dtb \
+	fsl-lx2160a-cex7-8-x-x.dtb \
+	fsl-lx2160a-half-twins-8-9-x.dtb \
 	fsl-lx2162a-som.dtb
 dtb-$(CONFIG_FSL_LSCH2) += fsl-ls1043a-qds-duart.dtb \
 	fsl-ls1043a-qds-lpuart.dtb \
diff --git a/arch/arm/dts/fsl-lx2160a-cex7-8-x-x.dts b/arch/arm/dts/fsl-lx2160a-cex7-8-x-x.dts
new file mode 100644
index 000000000..0cc95fa2d
--- /dev/null
+++ b/arch/arm/dts/fsl-lx2160a-cex7-8-x-x.dts
@@ -0,0 +1,196 @@
+// SPDX-License-Identifier: GPL-2.0+ OR X11
+/*
+ * SolidRun LX2160ACEX7 device tree source
+ *
+ * Author:	Rabeeh Khoury <rabeeh@solid-run.com>
+ *
+ * Copyright 2019 SolidRun ltd.
+ *
+ */
+
+/dts-v1/;
+
+#include "fsl-lx2160a.dtsi"
+
+/ {
+	model = "SolidRun LX2160ACEX7 COM express type 7 based board";
+	compatible = "fsl,lx2160acex7", "fsl,lx2160a";
+
+	aliases {
+		spi0 = &fspi;
+	};
+};
+
+&dpmac3 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac4 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+
+&dpmac5 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac6 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac7 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac8 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac9 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac10 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac17 {
+	status = "okay";
+	phy-handle = <&rgmii_phy1>;
+	phy-connection-type = "rgmii-id";
+};
+
+&emdio1 {
+	status = "okay";
+
+	rgmii_phy1: ethernet-phy@1 {
+		/* AR8035 PHY - "compatible" property not strictly needed */
+		compatible = "ethernet-phy-id004d.d072";
+		reg = <0x1>;
+		/* Poll mode - no "interrupts" property defined */
+	};
+};
+
+&fspi {
+	bus-num = <0>;
+	status = "okay";
+
+	qflash0: MT35XU512ABA1G12@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "spi-flash";
+		spi-max-frequency = <50000000>;
+		reg = <0>;
+		/* The following setting enables 1-1-8 (CMD-ADDR-DATA) mode */
+		fspi-rx-bus-width = <8>; /* 8 FSPI Rx lines */
+		fspi-tx-bus-width = <1>; /* 1 FSPI Tx line */
+	};
+
+};
+
+&esdhc0 {
+	status = "okay";
+};
+
+&esdhc1 {
+	status = "okay";
+};
+
+&i2c0 {
+	status = "okay";
+	u-boot,dm-pre-reloc;
+	i2c-mux@77 {
+		compatible = "nxp,pca9547";
+		reg = <0x77>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		i2c@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x0>;
+/* The following eeprin is reserved so that the 2Kb eeprom at address 0x57
+   would be used as TLV eeprom.
+			24aa512@50 {
+				compatible = "atmel,24c512";
+				reg = <0x50>;
+			};
+*/
+			spd1@51 {
+				compatible = "atmel,spd";
+				reg = <0x51>;
+			};
+			spd2@53 {
+				compatible = "atmel,spd";
+				reg = <0x53>;
+			};
+			m24c02@57 {
+				compatible = "atmel,24c02";
+				reg = <0x57>;
+			};
+		};
+		i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x1>;
+			fan-temperature-ctrlr@18 {
+				compatible = "ti,amc6821";
+				reg = <0x18>;
+				cooling-min-state = <0>;
+				cooling-max-state = <9>;
+				#cooling-cells = <2>;
+			};
+		};
+		i2c@2 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x2>;
+			ltc3882@5c {
+				compatible = "ltc3882";
+				reg = <0x5c>;
+			};
+		};
+		i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x3>;
+
+			temperature-sensor@48 {
+				compatible = "nxp,sa56004";
+				reg = <0x48>;
+			};
+		};
+	};
+};
+
+&i2c4 {
+	status = "okay";
+
+	rtc@51 {
+		compatible = "pcf2127-rtc";
+		reg = <0x51>;
+	};
+};
+
+&sata0 {
+	status = "okay";
+};
+
+&sata1 {
+	status = "okay";
+};
+
+&sata2 {
+	status = "okay";
+};
+
+&sata3 {
+	status = "okay";
+};
diff --git a/arch/arm/dts/fsl-lx2160a-half-twins-8-9-x.dts b/arch/arm/dts/fsl-lx2160a-half-twins-8-9-x.dts
new file mode 100644
index 000000000..5672e7368
--- /dev/null
+++ b/arch/arm/dts/fsl-lx2160a-half-twins-8-9-x.dts
@@ -0,0 +1,215 @@
+// SPDX-License-Identifier: GPL-2.0+ OR X11
+/*
+ * SolidRun LX2160ACEX7 device tree source
+ *
+ * Author:	Rabeeh Khoury <rabeeh@solid-run.com>
+ *
+ * Copyright 2019 SolidRun ltd.
+ *
+ */
+
+/dts-v1/;
+
+#include "fsl-lx2160a.dtsi"
+
+/ {
+	model = "SolidRun Half Twins Board";
+	compatible = "fsl,lx2160acex7", "fsl,lx2160a";
+
+	aliases {
+		spi0 = &fspi;
+	};
+};
+
+&dpmac3 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac4 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac5 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac6 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac7 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac8 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac9 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac10 {
+	status = "okay";
+	phy-connection-type = "xgmii";
+};
+
+&dpmac11 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&dpmac12 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&dpmac13 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&dpmac14 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&dpmac15 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&dpmac16 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&dpmac17 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&dpmac18 {
+	status = "okay";
+	phy-connection-type = "sgmii";
+};
+
+&fspi {
+	bus-num = <0>;
+	status = "okay";
+
+	qflash0: MT35XU512ABA1G12@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "spi-flash";
+		spi-max-frequency = <50000000>;
+		reg = <0>;
+		/* The following setting enables 1-1-8 (CMD-ADDR-DATA) mode */
+		fspi-rx-bus-width = <8>; /* 8 FSPI Rx lines */
+		fspi-tx-bus-width = <1>; /* 1 FSPI Tx line */
+	};
+
+};
+
+&esdhc0 {
+	status = "okay";
+};
+
+&esdhc1 {
+	status = "okay";
+};
+
+&i2c0 {
+	status = "okay";
+	u-boot,dm-pre-reloc;
+	i2c-mux@77 {
+		compatible = "nxp,pca9547";
+		reg = <0x77>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+		i2c@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x0>;
+/* The following eeprin is reserved so that the 2Kb eeprom at address 0x57
+   would be used as TLV eeprom.
+			24aa512@50 {
+				compatible = "atmel,24c512";
+				reg = <0x50>;
+			};
+*/
+			spd1@51 {
+				compatible = "atmel,spd";
+				reg = <0x51>;
+			};
+			spd2@53 {
+				compatible = "atmel,spd";
+				reg = <0x53>;
+			};
+			m24c02@57 {
+				compatible = "atmel,24c02";
+				reg = <0x57>;
+			};
+		};
+		i2c@1 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x1>;
+			fan-temperature-ctrlr@18 {
+				compatible = "ti,amc6821";
+				reg = <0x18>;
+				cooling-min-state = <0>;
+				cooling-max-state = <9>;
+				#cooling-cells = <2>;
+			};
+		};
+		i2c@2 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x2>;
+			ltc3882@5c {
+				compatible = "ltc3882";
+				reg = <0x5c>;
+			};
+		};
+		i2c@3 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x3>;
+
+			temperature-sensor@48 {
+				compatible = "nxp,sa56004";
+				reg = <0x48>;
+			};
+		};
+	};
+};
+
+&i2c2 {
+	status = "okay";
+	expander2: pca9555@24 {
+		compatible = "nxp,pca9555";
+		pinctrl-names = "default";
+		gpio-controller;
+		#gpio-cells = <2>;
+		interrupt-controller;
+		#interrupt-cells = <2>;
+		reg = <0x24>;
+	};
+};
+
+&i2c4 {
+	status = "okay";
+
+	rtc@51 {
+		compatible = "pcf2127-rtc";
+		reg = <0x51>;
+	};
+};
diff --git a/board/solidrun/lx2160a/eth_lx2160acex7.c b/board/solidrun/lx2160a/eth_lx2160acex7.c
index abca006d2..6a696deaf 100644
--- a/board/solidrun/lx2160a/eth_lx2160acex7.c
+++ b/board/solidrun/lx2160a/eth_lx2160acex7.c
@@ -223,6 +223,7 @@ static struct serdes_configuration {
 	{2, 2, false},
 	{2, 3, false},
 	{2, 5, false},
+	{2, 9, true},
 	{2, 10, false},
 	{2, 11, true},
 	{2, 12, true},
@@ -265,11 +266,33 @@ static void get_str_protocol(u8 serdes_block, u32 protocol, char *str)
 
 int board_fit_config_name_match(const char *name)
 {
+	struct ccsr_gur *gur = (void *)(CONFIG_SYS_FSL_GUTS_ADDR);
+	u32 srds_s1, srds_s2;
+	u32 rcw_status = in_le32(&gur->rcwsr[28]);
+
+	srds_s1 = rcw_status & FSL_CHASSIS3_RCWSR28_SRDS1_PRTCL_MASK;
+	srds_s1 >>= FSL_CHASSIS3_RCWSR28_SRDS1_PRTCL_SHIFT;
+
+	srds_s2 = rcw_status & FSL_CHASSIS3_RCWSR28_SRDS2_PRTCL_MASK;
+	srds_s2 >>= FSL_CHASSIS3_RCWSR28_SRDS2_PRTCL_SHIFT;
 	if (get_svr() & 0x800) {
 		if (!strncmp(name, "fsl-lx2162a-som", 15)) {
 			return 0;
 		}
 	} else {
+		/* Notice - order of matching is similar to order of CONFIG_OF_LIST */
+		if (srds_s1 == 8) {
+			if (srds_s2 == 9) { /* Half-twins board */
+				if (!strncmp(name, "fsl-lx2160a-half-twins-8-9-x", 28)) {
+					return 0;
+				}
+			}
+			if (srds_s2 <= 5) { /* No network with SD2 <= 5 */
+				if (!strncmp(name, "fsl-lx2160a-cex7-8-x-x", 22)) {
+					return 0;
+				}
+			}
+		}
 		if (!strncmp(name, "fsl-lx2160a-cex7", 16)) return 0;
 	}
 
@@ -350,7 +373,9 @@ int fsl_board_late_init(void) {
 			printf("SerDes1 protocol 0x%x is not supported on LX2160ACEX7\n",
 			       srds_s1);
 		}
-		sprintf(expected_dts, "fsl-lx2160a-clearfog-cx.dtb");
+		if ((srds_s1 == 8) && ((srds_s2 == 9) || (srds_s2 == 2))) {
+			sprintf(expected_dts, "fsl-lx2160a-half-twins.dtb");
+		} else sprintf(expected_dts, "fsl-lx2160a-clearfog-cx.dtb");
 	}
 	env_set("fdtfile", expected_dts);
 	return 0;
diff --git a/configs/lx2160acex7_tfa_defconfig b/configs/lx2160acex7_tfa_defconfig
index ba3568646..3dd227251 100644
--- a/configs/lx2160acex7_tfa_defconfig
+++ b/configs/lx2160acex7_tfa_defconfig
@@ -40,7 +40,7 @@ CONFIG_CMD_NVME=y
 CONFIG_NVME=y
 CONFIG_MP=y
 CONFIG_OF_CONTROL=y
-CONFIG_OF_LIST="fsl-lx2160a-cex7 fsl-lx2162a-som"
+CONFIG_OF_LIST="fsl-lx2160a-half-twins-8-9-x fsl-lx2160a-cex7-8-x-x fsl-lx2160a-cex7 fsl-lx2162a-som"
 CONFIG_MULTI_DTB_FIT=y
 CONFIG_ENV_OVERWRITE=y
 CONFIG_ENV_IS_IN_MMC=y
-- 
2.25.1

